homeassistant/components/zwave_js/discovery_data_template.py

```diff
@@ -251,3 +251,111 @@ def values_to_watch(self, resolved_data: dict[str, Any]) -> Iterable[ZwaveValue]
     def current_tilt_value(resolved_data: dict[str, Any]) -> ZwaveValue | None:
         """Get current tilt ZwaveValue from resolved data."""
         return resolved_data["tilt_value"]
+
+
+@dataclass
+class FanSpeedDataTemplate:
+    """Mixin to define get_speed_config."""
+
+    def get_speed_config(self, resolved_data: dict[str, Any]) -> list[int] | None:
+        """
+        Get the fan speed configuration for this device.
+
+        Values should indicate the highest allowed device setting for each
+        actual speed.
+        """
+        # pylint: disable=no-self-use
+        raise NotImplementedError
+
+
+@dataclass
+class FixedFanSpeedValueMix:
+    """Mixin data class for defining supported fan speeds."""
+
+    speeds: list[int]
+
+
+@dataclass
+class FixedFanSpeedDataTemplate(
+    BaseDiscoverySchemaDataTemplate, FanSpeedDataTemplate, FixedFanSpeedValueMix
+):
+    """
+    Specifies a fixed set of fan speeds.
+
+    Example:
+      ZWaveDiscoverySchema(
+          platform="fan",
+          hint="configured_fan_speed",
+          ...
+          data_template=FixedFanSpeedDataTemplate(
+              speeds=[32,65,99]
+          ),
+      ),
+
+    `speeds` indicates the maximum setting on the underlying fan controller
+    for each actual speed.
+    """
+
+    def get_speed_config(self, resolved_data: dict[str, Any]) -> list[int] | None:
+        """Get the fan speed configuration for this device."""
+        return self.speeds
+
+
+@dataclass
+class ConfigurableFanSpeedValueMix:
+    """Mixin data class for defining configurable fan speeds."""
+
+    configuration_option: ZwaveValueID
+    configuration_value_to_speeds: dict[int, list[int]]
+
+
+@dataclass
+class ConfigurableFanSpeedDataTemplate(
+    BaseDiscoverySchemaDataTemplate, FanSpeedDataTemplate, ConfigurableFanSpeedValueMix
+):
+    """
+    Gets fan speeds based on a configuration value.
+
+    Example:
+      ZWaveDiscoverySchema(
+          platform="fan",
+          hint="configured_fan_speed",
+          ...
+          data_template=ConfigurableFanSpeedDataTemplate(
+            configuration_option=ZwaveValueID(
+                5, CommandClass.CONFIGURATION, endpoint=0
+            ),
+            configuration_value_to_speeds={0: [32, 65, 99], 1: [24, 49, 74, 99]},
+          ),
+      ),
+
+    `configuration_option` is a reference to the setting that determines how
+    many speeds are supported.
+
+    `configuration_value_to_speeds` maps the values from `configuration_option`
+    to a list of speeds.  The specified speeds indicate the maximum setting on
+    the underlying switch for each actual speed.
+    """
+
+    def resolve_data(self, value: ZwaveValue) -> dict[str, Any]:
```

Comment:
We should probably type the value as `ConfigurationValue`.
https://github.com/home-assistant-libs/zwave-js-server-python/blob/e7b7c5327d982cd3bcad973c96fca516026df496/zwave_js_server/model/value.py#L278

Reply:
Hrm, it might not be legit to narrow down the type of the input parameter here, since BaseDiscoverySchemaDataTemplate defines it as taking 'ZwaveValue' (LSP and all).  I could narrow down the return value though.

Reply:
Updated the return type.